tx_test.go

```diff
@@ -95,3 +95,31 @@ func TestTxErrors(t *testing.T) {
 			"got %v, want %v", err, io.EOF)
 	}
 }
+
+// TestTxWitness tests the HasWitness() and WitnessHash() methods.
+func TestTxWitness(t *testing.T) {
+	msgTxWithWitness := Block100000.Transactions[0] // contains witness data
+	txWithWitness := btcutil.NewTx(msgTxWithWitness)
+
+	txWithWitness.WitnessHash() // Populate the cached witness hash
+	txWithWitness.HasWitness()  // Should not fail (see btcsuite/btcd#1543)
+
+	if !txWithWitness.HasWitness() {
+		t.Errorf("HasWitness: got false, want true")
+	}
+	if txWithWitness.WitnessHash().IsEqual(txWithWitness.Hash()) {
+		t.Errorf("WitnessHash: witness hash and tx id must NOT be same - "+
+			"got %v, want %v", txWithWitness.WitnessHash(), txWithWitness.Hash())
+	}
+
+	for _, msgTxWithoutWitness := range Block100000.Transactions[1:] {
+		txWithoutWitness := btcutil.NewTx(msgTxWithoutWitness)
+		if txWithoutWitness.HasWitness() {
+			t.Errorf("HasWitness: got false, want true")
+		}
+		if !txWithoutWitness.WitnessHash().IsEqual(txWithoutWitness.Hash()) {
+			t.Errorf("WitnessHash: witness hash and tx id must be same - "+
+				"got %v, want %v", txWithoutWitness.WitnessHash(), txWithoutWitness.Hash())
+		}
+	}
+}
```

Comment:
While it's great that you're testing both methods, it's usually more idiomatic to test each method independently. Then, in regression testing, if `TestTxHasWitness` breaks and a new test, `TestTxHashWitness`, does not, then we more or less know the bug is in `TestTxHasWitness` and can focus on that, rather than trying to figure out which method is incorrect after seeing that only `TestTxWitness` fails.

Reply:
Done! Although, I wanted to include a proof-of-fix in my PR, so in the test for `HasWitness()`, I'm calling the `WitnessHash()` first, since that's how the bug is reproduced. Let me know if this is acceptable.

Reply:
I think that makes sense, looks good!